src/macaron/parsers/bashparser.py

```diff
@@ -115,78 +148,95 @@ def parse(bash_content: str, macaron_path: str = "") -> dict:
+      * :class:`BashScriptType.FILE` when it is a bash script file.
 
     Parameters
     ----------
-    bash_content : str
-        Bash script content.
-    ci_file : str
-        The relative path to the entry point CI workflow.
-    ci_type : str
-        The CI service.
-    macaron_path : str
-        Macaron's root path (optional).
-    recursive : bool
-        Recursively parse bash scripts with depth 1.
-        Should specify repo_path too if set to True.
-    repo_path : str
+    name: str
+        A name to be used as the identifier of the node.
+    node_type: BashScriptType
+        The type of the node.
+    source_path: str
+        The file that contains the bash script.
+    parsed_obj: dict | None
+        The parsed bash scrit object.
+    repo_path: str
         The path to the target repo.
-    working_dir : str
-        The working directory from which the bash script has run.
-        Empty value is considered as the root of the repo.
-    job_name: str
-        The name of the job where commands were called.
-    step_name: str
-        The name of the step where commands were called.
-
-    Yields
+    caller: BaseNode
+        The caller node.
+    recursion_depth: int
+        The number of times this function is called recursively.
+    macaron_path=None
+        The path to the Macaron module.
+
+    Returns
+    -------
+    BashNode
+        A bash node object.
+
+    Raises
     ------
-    BashCommands
-        The parsed bash script objects.
+    CallGraphError
+        When unable to create a bash node.
     """
-    if not macaron_path:
-        macaron_path = global_config.macaron_path
-
-    parsed_parent = parse(bash_content)
-    caller_commands = parsed_parent.get("commands", [])
-    if caller_commands:
-        yield BashCommands(
-            caller_path=ci_file,
-            CI_path=ci_file,
-            CI_type=ci_type,
-            commands=caller_commands,
-            job_name=job_name,
-            step_name=step_name,
-        )
+    if recursion_depth > defaults.getint("bashparser", "recursion_depth", fallback=3):
+        raise CallGraphError(f"The analysis has reached maximum recursion depth {recursion_depth} at {source_path}.")
+    parsed_content = {}
+    working_dir = None
+    match node_type:
+        case BashScriptType.INLINE:
+            if parsed_obj is None:
+                raise CallGraphError(f"Unable to find the parsed AST for the CI step at {source_path}.")
+            step_exec = validate_step(parsed_obj)
+            if step_exec is None:
+                raise CallGraphError(f"Unable to validate parsed AST for the CI step at {source_path}.")
+
+            working_dir = step_exec.get("WorkingDirectory")
+            run_script = validate_run_step(parsed_obj)
+            if run_script is None:
+                raise CallGraphError(f"Invalid run step at {source_path}.")
+            parsed_content = parse(run_script, macaron_path=macaron_path)
```

Comment:
From my understanding. The non-recursive entry to this `create_bash_node` function, `node_type` will always be `BashScriptType.INLINE`. This is because we call this function whenever we encounter an inline bash run step in GitHub Actions workflow
https://github.com/oracle/macaron/blob/ed96daea0f0c4ae81f6f42b464d859f0829563aa/src/macaron/slsa_analyzer/ci_service/github_actions/analyzer.py#L261-L287

Therefore, match case for `case BashScriptType.INLINE` will only run once. The recursive calls on `create_bash_node` will only happen with `node_type` as `BashScriptType.FILE`.

In additions, the logic for this `case BashScriptType.INLINE`  block contains validation for GitHub Action parsed obj (e.g. `validate_run_step` and ` validate_step`). 

With those reasons, I wonder if we could refactor the validation of GitHub Action parsed obj outside of this function, which is only supposed to handle bash related parsing. 
I believe the best place where we could place validation GitHub Action parsed obj is before this try block.  
https://github.com/oracle/macaron/blob/ed96daea0f0c4ae81f6f42b464d859f0829563aa/src/macaron/slsa_analyzer/ci_service/github_actions/analyzer.py#L268

Overall the flow can be described as follows:
```
            step_exec = validate_step(step)
            if step_exec is None:
                raise CallGraphError(f"Unable to validate parsed AST for the CI step at {node.source_path}.")

             working_dir = step_exec.get("WorkingDirectory")
             run_script = validate_run_step(step_exec)
             if run_script is None:
                    raise CallGraphError(f"Invalid run step at {node.source_path}.")
             
             # The rest of the try block. 
```
This would also require us to change the signature of the `create_bash_node` function to accept the validated data from outside.

Reply:
I have a further question as well. Is there any case where a `BashScriptType.FILE` bash script calls a `BashScriptType.INLINE` bash script :thinking: ? 
If that is never the case, I believe we can separate the handling of `BashScriptType.FILE` out of this function too. What I am thinking is
```python
# within the build_call_graph_from_node function
if (
    step.get("Exec")
    and step["Exec"].get("Run")
    and (step["Exec"].get("Shell") is None or step["Exec"]["Shell"].get("Value") in ["bash", "sh"])
):
    # validate run step parsed obj
    ....

    # create inline bash node
    try:
        inline_bash_node = create_inline_bash_node(...)
    except CallGraphError as error:
        # handle error

    # create callee file bash nodes
    try:
        callee_bash_nodes: list[BashNode]  = create_callee_bash_nodes(...)
   except CallGraphError as error:
       # handle error

   for callee in callee_bash_nodes:
       inline_bash_node.add_callee(callee)
        
   job_node.add_callee(inline_bash_node)
```

Reply:
I'm not sure about the benefit of suggested refactorings. I think adding a separate function will require repeating code, whereas the current implementation takes the node type as argument and picks the right logic accordingly.